lib/sidekiq/ultimate/resurrector/lock.rb

```diff
@@ -20,20 +20,32 @@ class Lock
         class << self
           def acquire
             Sidekiq.redis do |redis|
-              Redlock::Client.new([redis]).lock(LOCK_KEY, LOCK_TTL) do |locked|
-                break unless locked
-
-                results  = redis.pipelined { |pipeline| [pipeline.time, pipeline.get(LAST_RUN_KEY)] }
-                distance = results[0][0] - results[1].to_i
+              break if resurrected_recently?(redis) # Cheap check since lock will not be free most of the time
 
-                break unless CommonConstants::RESURRECTOR_INTERVAL < distance
+              Redlock::Client.new([redis], :retry_count => 0).lock(namespaced_lock_key, LOCK_TTL) do |locked|
+                break unless locked
+                break if resurrected_recently?(redis)
 
                 yield
 
                 redis.set(LAST_RUN_KEY, redis.time.first)
               end
             end
           end
+
+          def namespaced_lock_key
+            return @namespaced_lock_key if defined?(@namespaced_lock_key)
+
+            namespace = Sidekiq.redis { |redis| redis.namespace if redis.respond_to?(:namespace) }
+            @namespaced_lock_key = "#{namespace}:#{LOCK_KEY}"
+          end
+
+          def resurrected_recently?(redis)
+            results  = redis.pipelined { |pipeline| [pipeline.time, pipeline.get(LAST_RUN_KEY)] }
+            distance = results[0][0] - results[1].to_i
```

Comment:
Can't you write something like

```suggestion
            start, end  = redis.pipelined { |pipeline| [pipeline.time, pipeline.get(LAST_RUN_KEY)] }
            distance = end[0] - start.to_i
```

Reply:
And why is only one of these `.to_i`?

Reply:
`TIME` is guaranteed to return two integers, `GET` on the other hand, can return just about anything. In this case we trust that `LAST_RUN_KEY` stores `Integer`, but it might not be set yet, in this case it will return `nil`

Reply:
BTW, there's a bit more fancy way to write this:

``` ruby
(curr, *), last = redis.pipelined { |pipeline| [pipeline.time, pipeline.get(LAST_RUN_KEY)] }
distance = curr - last.to_i
```

Furthermore, I think it's better to use `MULTI` instead of `PIPELINE`.